SHT3xD failed communication when I2C frequency >50-100kHz using esp-idf but works well over this frequency with arduino framework

### The problem

I am converting all my ESP nodes from Arduino to ESP-IDF and ran into an unexpected issue with the SHT3xD sensor. The issue is present on all my nodes that use that sensor and across multiple brands and implementations of the sensor thus likely excluding a hardware issue.

With the default Arduino framework, I can set the i2c frequency to 400khz (to eliminate warnings of a blocking component) and the SHT3xD sensor works perfectly. If I change the framework to `esp-idf`, the sensor fails to communicate unless the frequency is at the default 50khz, but in a few cases I have been able to set it to 100khz, even 150khz, but it fails at 200khz. The exact cutoff may depend on what else is happening on the i2c bus, or the combination of other devices on the bus. On the 2 nodes I used for testing, I verified the pullup resistors to be appropriate (measured both connected and disconnected).

I have verified this issue with all esphome releases going back to 2023.7 (latest for each month) and it persists.

### Which version of ESPHome has the issue?

2023.12.5

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

_No response_

### What platform are you using?

ESP32

### Board

_No response_

### Component causing the issue

SHT3xD / i2c

### Example YAML snippet

```yaml
This is a work in progress on a project I just started but easier to share than the node I discovered this issue with that has many more sensors, including an OLED display which was the main reason I really needed the 400khz i2c frequency. Either way, the issue shows up with this simpler node as well:


substitutions:
  devicename: downstairs-hvac
  devicename_no_dashes: downstairs_hvac
  friendly_devicename: "Downstairs HVAC"
  device_description: "Downstairs HVAC"
  update_interval_s: "20s"
  update_interval_wifi: "120s"
  temperature_calibration: "0.0"
  humidity_calibration: "0.0"


esphome:
  name: ${devicename}
  comment: ${device_description}
  friendly_name: ${friendly_devicename}

esp32:
  board: nodemcu-32s
  framework:
    type: arduino
    #type: esp-idf
    version: recommended

wifi:
  ssid: !secret iot_wifi_ssid
  password: !secret iot_wifi_password
  power_save_mode: none

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "${devicename}"
    password: !secret iot_wifi_password

#Faster than DHCP. Also use if can't reach because of name change
  manual_ip:
    static_ip: 10.1.3.202
    gateway: 10.1.2.1
    subnet: 255.255.254.0
    dns1: 10.1.0.50
    dns2: 10.1.0.51

#Manually override what address to use to connect to the ESP.
#Defaults to auto-generated value. Example, if you have changed your
#static IP and want to flash OTA to the previously configured IP address.
  use_address: 10.1.3.202
  
logger:
#    baud_rate: 0 #disabled

# Enable Home Assistant API
api:
  encryption:
    key: "redacted"
  
ota:
  
web_server:
  port: 80
  include_internal: true

# Sync time with Home Assistant
time:
  - platform: homeassistant
    id: ha_time

i2c:
  - id: bus_a
    sda: 21
    scl: 22
    scan: true
    frequency: 400khz

#Configuration entry for 18B20 sensor
dallas:
  - pin: 27
    update_interval: ${update_interval_s}

text_sensor:
  - platform: wifi_info
    ip_address:
      name: "IP"
      icon: "mdi:ip-outline"
      update_interval: ${update_interval_wifi}
    ssid:
      name: "SSID"
      icon: "mdi:wifi-settings"
      update_interval: ${update_interval_wifi}
    bssid:
      name: "BSSID"
      icon: "mdi:wifi-settings"
      update_interval: ${update_interval_wifi}
    mac_address:
      name: "MAC"
      icon: "mdi:network-outline"
    scan_results:
      name: "Wifi Scan"
      icon: "mdi:wifi-refresh"
      disabled_by_default: true
  
sensor:
  - platform: wifi_signal
    name: "WiFi Signal"
    update_interval: ${update_interval_wifi}
    device_class: signal_strength

  - platform: sht3xd
    i2c_id: bus_a
    address: 0x45
    temperature:
      id: temp
      name: "Temperature"
      filters:
        - lambda: return x + ${temperature_calibration};
    humidity:
      id: humidity
      name: "Humidity"
      filters:
        - lambda: return x + ${humidity_calibration};
    update_interval: ${update_interval_s}

  - platform: dallas
    address: 0xd9000000176c9b28
    name: "Return Air Temperature"
    id: return_air
    accuracy_decimals: 1

  - platform: dallas
    address: 0x6600000016695728
    name: "Supply Air Temperature"
    id: supply_air
    accuracy_decimals: 1

  - platform: template
    name: "Air Temperature Delta"
    id: air_temperature_delta
    lambda: 'return (id(supply_air).state - id(return_air).state);'
    device_class: temperature
    unit_of_measurement: "°C"
    state_class: "measurement"
    accuracy_decimals: 1

binary_sensor:
  - platform: status
    name: "Connection Status"
    id: connection_status
    entity_category: diagnostic

  - platform: gpio
    pin:
      number: 13
      inverted: true
      mode:
        input: true
        pullup: true
    name: "Water Sensor"
    id: water_sensor


switch:
  - platform: restart
    name: "Restart"

button:
  - platform: safe_mode
    name: "Restart (Safe Mode)"
```
```


### Anything in the logs that might be useful for us?

_No response_

### Additional information

_No response_

@mrtoy-me Thanks for the work on this!

> I made a few changes to the test sht3xd component in https://github.com/mrtoy-me/esphome-components-test

FYI this version throws compilation errors:
```
src/esphome/components/sht3xd/sht3xd.cpp: In member function 'virtual void esphome::sht3xd::SHT3XDComponent::setup()':
src/esphome/components/sht3xd/sht3xd.cpp:31:11: error: 'class esphome::sht3xd::SHT3XDComponent' has no member named 'error_code_'
   31 |     this->error_code_ = WRITE_BREAK_FAILED;
      |           ^~~~~~~~~~~
src/esphome/components/sht3xd/sht3xd.cpp:31:25: error: 'WRITE_BREAK_FAILED' was not declared in this scope; did you mean 'WRITE_CLEAR_FAILED'?
   31 |     this->error_code_ = WRITE_BREAK_FAILED;
      |                         ^~~~~~~~~~~~~~~~~~
      |                         WRITE_CLEAR_FAILED
src/esphome/components/sht3xd/sht3xd.cpp:38:11: error: 'class esphome::sht3xd::SHT3XDComponent' has no member named 'error_code_'
   38 |     this->error_code_ = WRITE_SOFT_RESET_FAILED;
      |           ^~~~~~~~~~~
src/esphome/components/sht3xd/sht3xd.cpp:38:25: error: 'WRITE_SOFT_RESET_FAILED' was not declared in this scope
   38 |     this->error_code_ = WRITE_SOFT_RESET_FAILED;
      |                         ^~~~~~~~~~~~~~~~~~~~~~~
*** [.pioenvs/airgradient-downstairs/src/esphome/components/sht3xd/sht3xd.cpp.o] Error 1
```

Reverting to the last commit works. I haven't tried to pull the component down and debug the code yet.

For anyone else - here's the external_components block with the explicit reference to the commit.
```
external_components:
   - source:
        type: git
        url: https://github.com/mrtoy-me/esphome-components-test
        ref: 14aeb0781eed158d6c1c58b39c4ceabd6a81797a
     components: [ sht3xd ]
```

--- a/components/sensor/sht3xd.rst
+++ b/components/sensor/sht3xd.rst
@@ -54,6 +54,24 @@ Configuration variables:
   but can also increase temperature readings and decrease humidity readings as a side effect.
   Defaults to ``false``.
 
+Configuration using Higher I²C Frequencies
+------------------------------------------
+
+When using the **IDF framework** and **I²C Frequencies greater than 50-100kHz**, the I²C configuration needs to include a **timeout** option.
+On an ESP32, the Arduino framework has a default I²C timeout of 50ms whereas on IDF framework, the default timeout is 100us.
+At these higher I²C Frequencies, the default I²C timeout on IDF framework causes a "Communication with SHT3xD failed" error on setup.
+A solution that has been tested on ESP32 at 800kHz, is to increase the I²C timeout to 10ms as per the following example.
+
+.. code-block:: yaml
+
+    # Example I²C configuration entry
+    i2c:
+      sda: 21
+      scl: 22
+      scan: true
+      frequency: 800khz
+      timeout: 10ms
+
 See Also
 --------
 

--- a/components/sensor/sht3xd.rst
+++ b/components/sensor/sht3xd.rst
@@ -54,17 +54,17 @@ Configuration variables:
   but can also increase temperature readings and decrease humidity readings as a side effect.
   Defaults to ``false``.
 
-Configuration using Higher I²C Frequencies
-------------------------------------------
+I²C Configuration when using Higher I²C Frequencies
+---------------------------------------------------
 
-When using the **IDF framework** and **I²C Frequencies greater than 50-100kHz**, the I²C configuration needs to include a **timeout** option.
+When using the **IDF framework** and **I²C frequencies greater than 50-100kHz**, the I²C configuration needs to include a **timeout** option.
 On an ESP32, the Arduino framework has a default I²C timeout of 50ms whereas on IDF framework, the default timeout is 100us.
-At these higher I²C Frequencies, the default I²C timeout on IDF framework causes a "Communication with SHT3xD failed" error on setup.
+At these higher I²C frequencies, the default I²C timeout on IDF framework causes a "Communication with SHT3xD failed" error on setup.
 A solution that has been tested on ESP32 at 800kHz, is to increase the I²C timeout to 10ms as per the following example.
 
 .. code-block:: yaml
 
-    # Example I²C configuration entry
+    # Example I²C configuration
     i2c:
       sda: 21
       scl: 22

